### DISCUSSION TITLE
🚀[FEATURE]: Include PropertyPath to NGXS Form Plugin Directive

### DISCUSSION THREAD

**Discussion by jermylucas** [ngxs/store#1905] [Ideas]
### Relevant Package
@ngxs/form-plugin

### Description
NGXS Forms already have the `propertyPath` parameter that can be used on the form Update actions. I would like to request that this is added to the `ngxsForm` directive. This may resolve a few potential problems with the current form plugin, as well as make things slightly more flexible for those occasions where the directive is used.

### Describe the problem you are trying to solve
If I have a multi-step form split up into multiple components, where I want each component to manage the smaller form group, but I want all the values updated on the state, I am unable to break things up and using the `ngxsForm` directive does not allow this to be done properly. For example, say in my NGXS state I have:
```
export interface ProductForm {
  model: ProductFormModel;
  dirty: false;
  status: '';
  errors: {};
}

export interface ProductFormModel {
  customerInfo: Customer;
  product: Product[];
  checkoutInfo: Checkout;
}
```

But I have the steps broken out into three components: `customer.component`, `product.component`, and `checkout.component`. For validation, I want to simply use a form group on each component, but update the entire form value in the store. If I were to use `ngxsForm="...productForm"` on each component, the values in the model would be reset each time the page changes, and they would not be nested in objects as intended through the interface. If I wanted to do something like `ngxsForm="...productForm.customerInfo"`, then of course a new path and model is created.

Angular also allows nested form groups. If the same above example was done with a "main" form group, like:
```
productForm = this.fb.group({
  customerInfo: this.fb.group({...}),
  productInfo: this.fb.group({...}),
  checkoutInfo: this.fb.group({...}),
})
```
and then I had my HTML as:
```
<form [formGroup]"productForm">
  <router-outlet></router-outlet>
</form>
```
Where each component in the outlet has the respective form groups, then ngxs would trigger an infinite loop of updates. 

I _believe_ this would fix these problems, and possibly a few others.

### Describe the solution you'd like
There could be an optional property added to the directive, something like `ngxsFormPropertyPath`, where the directive will take this into account while doing the updates on form changes. Or, it could be added to the directive as it already is. For the above examples, maybe something like: `ngxsForm="...productForm['customerInfo']"` where it is an optional value to the path, and the directive is able to use it in the updates. Currently is is like so in the `directive.ts`, but with changes, the property path would be here too:
```
new UpdateFormValue({
  path: this.path,
  value
}),
new UpdateFormDirty({
  path: this.path,
  dirty: this._formGroupDirective.dirty
}),
new UpdateFormErrors({
  path: this.path,
  errors: this._formGroupDirective.errors
})
```

I believe this is not too complex of a feature, since `propertyPath` is already valid in NGXS. 

I do not know if this solves the nested form group issue, or issues like #1822 . However, it gives more flexibility to use propertyPath in the above circumstances, as well as form arrays, to update groups inside the form rather than changing or resetting the entire form property. 

### Describe alternatives you've considered
I am currently using a simple workaround. With the multi-component example, I am subscribing to `valueChanges` of the form, and updating the state with the property path every time something on the form changes:

```
this.customerInfoForm.valueChanges.subscribe((res) => {
  this.store.dispatch(
    new UpdateFormValue({
      path: 'product.productForm',
      value: {
        ...res,
      },
      propertyPath: 'customerInfo',
    })
  );
});
```
This is working as expected. I could add the `UpdateFormStatus` and `UpdateFormDirty` too, but it seems like a built in solution to the directive would also make sense.

**Comment by markwhitfeld** [comment]
@jermylucas Interesting suggestion!
I think that the `ngxsFormPropertyPath` sounds like a good option.
The one thing that is unclear, is, what would the `dirty`, `status` and `errors` properties be synchronised/represented if they are representing multiple forms? When one of the sub-property forms is dirty, or in an error state, does that mean that the others are too? Just trying to wrap my head around how that would play out...

Unless you are saying that they would all be under the same `ngForm` parent component. Then it should just work without the sub-property 🤔 .

The alternative to this is to have these 3 forms target 3 different properties in the state and therefore they will have their own form `status, etc.` props. That is ok. The data in the state does not need to look perfect. You can always compose the final object you are looking for with a selector. This is generally my recommended approach.

**Comment by jermylucas** [comment]
@markwhitfeld Thanks for the reply and for the great question. That's a very valid point. When I created the issue, I had assumed that the `propertyPath` was a parameter on more than just the `UpdateFormValue` action, but looking now it seems that this would only be an action to update the state values, and not the status of the forms. 

To answer your question though, I would assume that the `dirty`, `status`, and `errors` would be synchronized with the "parent" form in the state. Without the sub-property, the behavior on these is working correctly, but the nested values are not being set correctly (it resets previous form state values to patch in new values on each form unless parent form is the same form as is used in children). However, updating the status of the parent doesn't really make sense, because the state of said parent form would (and should) be different if the form object is broken apart across multiple components, each with their own unique validation and statuses. So unless each property path contained its own set of `dirty`, `status`, and `errors`, then this is more of a non-issue. 

I saw a clear simplification of the form values in the state, but it seems like:
- If values are all I care about, I should just stick with my current `valueChanges` approach
- If I care about the values, and status of forms, instead of trying to use them in the `propertyPath` I should instead break up the state into smaller forms per-page as well as per your latter suggestion
